lib/Runtime/PlatformAgnostic/Platform/Common/Intl.cpp

```diff
@@ -654,10 +675,210 @@ namespace Intl
+            -1,
+            reinterpret_cast<const UChar *>(pattern),
+            -1,
+            &status
+        );
+        IPlatformAgnosticResource *formatterResource = new IcuCObject<UDateFormat>(dtf, &udat_close);
+        AssertOrFailFast(formatterResource);
+        *resource = formatterResource;
+    }
+
+    // Formats `date` using the UDateFormat wrapped by `resource`. If `formatted` is non-null, the result is written into `formatted`
+    // Returns the number of characters written into `formatted` (including a null terminator) [should always be positive]
+    int FormatDateTime(_In_ IPlatformAgnosticResource *resource, _In_ double date, _Out_writes_opt_(formattedLen) char16 *formatted, _In_ int formattedLen)
+    {
+        UErrorCode status = U_ZERO_ERROR;
+        UDateFormat *dtf = UNWRAP_COBJECT(resource, UDateFormat);
+
+        int required = udat_format(dtf, date, reinterpret_cast<UChar *>(formatted), formattedLen, nullptr, &status);
+        if (formatted == nullptr && formattedLen == 0 && status == U_BUFFER_OVERFLOW_ERROR)
+        {
+            // when we are just counting bytes, we can ignore errors
+            AssertOrFailFast(required > 0);
+        }
+        else
+        {
+            ICU_ASSERT(status, required > 0 && required < formattedLen);
+        }
+
+        return required + 1;
+    }
+
+    // Formats `date` using the UDateFormat wrapped by `resource`. If `formatted` is non-null, the result is written into `formatted`
+    // Takes an additional parameter, fieldIterator, which is set to a IPlatformAgnosticResource* that should later be given to GetDateTimePartInfo
+    // Returns the number of characters written into `formatted` (includes a null terminator) [should always be positive]
+    int FormatDateTimeToParts(_In_ IPlatformAgnosticResource *resource, _In_ double date, _Out_writes_opt_(formattedLen) char16 *formatted,
+        _In_ int formattedLen, _Out_opt_ IPlatformAgnosticResource **fieldIterator)
+    {
+        UErrorCode status = U_ZERO_ERROR;
+        UDateFormat *dtf = UNWRAP_COBJECT(resource, UDateFormat);
+
+        UFieldPositionIterator *fpi = nullptr;
+
+        if (fieldIterator)
+        {
+            fpi = ufieldpositer_open(&status);
+            ICU_ASSERT(status, true);
+            IPlatformAgnosticResource *fpiResource = new IcuCObject<UFieldPositionIterator>(fpi, &ufieldpositer_close);
+            AssertOrFailFastMsg(fpiResource, "Out of memory");
+            *fieldIterator = fpiResource;
+        }
+
+        int required = udat_formatForFields(dtf, date, reinterpret_cast<UChar *>(formatted), formattedLen, fpi, &status);
+        if (formatted == nullptr && formattedLen == 0 && status == U_BUFFER_OVERFLOW_ERROR)
+        {
+            // when we are just counting bytes, we can ignore errors
+            AssertOrFailFast(required > 0);
+        }
+        else
+        {
+            ICU_ASSERT(status, required > 0 && required < formattedLen);
```

Comment:
>required > 0 && required < formattedLen [](start = 31, length = 39)

nit: conditions like this that show a var in the range between two other vars might be more readable as `0 < a && a < MAX` so that it looks more like math notation `0 < a < MAX`

Reply:
I didn't do this partly to avoid yoda-conditions and partly to match the other asserts that only want required > 0.